src/index/sparse/sparse_index_node.cc

```diff
@@ -0,0 +1,219 @@
+#include "index/sparse/sparse_inverted_index_config.h"
+#include "io/memory_io.h"
+#include "knowhere/comp/thread_pool.h"
+#include "knowhere/config.h"
+#include "knowhere/dataset.h"
+#include "knowhere/expected.h"
+#include "knowhere/factory.h"
+#include "knowhere/index_node.h"
+#include "knowhere/log.h"
+#include "knowhere/sparse_utils.h"
+#include "knowhere/utils.h"
+
+namespace knowhere {
+
+template <typename T, bool use_wand>
+class SparseInvertedIndexNode : public IndexNode {
+    static_assert(std::is_same_v<T, fp32>, "SparseInvertedIndexNode only support float");
+
+ public:
+    explicit SparseInvertedIndexNode(const int32_t& /*version*/, const Object& /*object*/)
+        : search_pool_(ThreadPool::GetGlobalSearchThreadPool()) {
+    }
+
+    ~SparseInvertedIndexNode() override {
+        if (index_ != nullptr) {
+            delete index_;
+            index_ = nullptr;
+        }
+    }
+
+    Status
+    Train(const DataSet& dataset, const Config& config) override {
+        auto cfg = static_cast<const SparseInvertedIndexConfig&>(config);
+        if (!IsMetricType(cfg.metric_type.value(), metric::IP)) {
+            LOG_KNOWHERE_ERROR_ << Type() << " only support metric_type: IP";
+            return Status::invalid_metric_type;
+        }
+        auto drop_ratio_build = cfg.drop_ratio_build.value_or(0.0f);
+        auto index = new sparse::InvertedIndex<T>();
+        index->SetUseWand(use_wand);
+        index->Train(static_cast<const sparse::SparseRow<T>*>(dataset.GetTensor()), dataset.GetRows(),
+                     drop_ratio_build);
+        if (index_ != nullptr) {
+            LOG_KNOWHERE_WARNING_ << Type() << " deleting old index during train";
+            delete index_;
+        }
+        index_ = index;
+        return Status::success;
+    }
+
+    Status
+    Add(const DataSet& dataset, const Config& config) override {
+        if (!index_) {
+            LOG_KNOWHERE_ERROR_ << "Could not add data to empty " << Type();
+            return Status::empty_index;
+        }
+        return index_->Add(static_cast<const sparse::SparseRow<T>*>(dataset.GetTensor()), dataset.GetRows(),
+                           dataset.GetDim());
+    }
+
+    [[nodiscard]] expected<DataSetPtr>
+    Search(const DataSet& dataset, const Config& config, const BitsetView& bitset) const override {
+        if (!index_) {
+            LOG_KNOWHERE_ERROR_ << "Could not search empty " << Type();
+            return expected<DataSetPtr>::Err(Status::empty_index, "index not loaded");
+        }
+        auto cfg = static_cast<const SparseInvertedIndexConfig&>(config);
+        auto nq = dataset.GetRows();
+        auto queries = static_cast<const sparse::SparseRow<T>*>(dataset.GetTensor());
+        auto k = cfg.k.value();
+        auto refine_factor = cfg.refine_factor.value_or(10);
+        auto drop_ratio_search = cfg.drop_ratio_search.value_or(0.0f);
+
+        auto p_id = new sparse::label_t[nq * k];
```

Comment:
same for using `std::make_unique<>` as above

Reply:
Done